File: LeetCode - Algorithm/Binary Tree/98.py
# 98. Validate Binary Search Tree
# https://leetcode.com/problems/validate-binary-search-tree/


from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.checkBST(root)

    def checkBST(
        self,
        node: Optional[TreeNode],
        smallerThan: Optional[int] = None,
        largerThan: Optional[int] = None,
    ) -> bool:
        if not node:
            return True

        if node.left and node.left.val >= node.val:
            return False

        if node.left and largerThan and node.left.val <= largerThan:
            return False

        if node.right and node.right.val <= node.val:
            return False

        if node.right and smallerThan and node.right.val >= smallerThan:
            return False

        return self.checkBST(
            node=node.left, smallerThan=node.val, largerThan=largerThan
        ) and self.checkBST(
            node=node.right, smallerThan=smallerThan, largerThan=node.val
        )

File: LeetCode - Algorithm/Binary Tree/98.js
// 98. Validate Binary Search Tree
// https://leetcode.com/problems/validate-binary-search-tree/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
  return checkBST(root, null, null);
};

const checkBST = (node, lowerLimit, upperLimit) => {
  let result = true;

  if (node.left) result = result && checkBST(node.left, lowerLimit, node.val);
  if (node.right) result = result && checkBST(node.right, node.val, upperLimit);

  if (lowerLimit !== null) result = result && lowerLimit < node.val;
  if (upperLimit !== null) result = result && node.val < upperLimit;

  return result;
};
